aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -4,13 +4,149 @@
+      sparse_csr::layoutToString(target_layout, false, true),
+      ": Expected product of batch dimensions to be non-zero.");
+
+  // Compute the number of non-zero elements in the first batch, expand to full
+  // size
+  auto nse_per_batch = mask.select(0, 0).sum().expand(mask.size(0));
+  TORCH_CHECK(
+      mask.sum({-2, -1}).equal(nse_per_batch),
+      "Expect the same number of specified elements per batch.");
+
+  // We need to join batches into a matrix increasing the length of the
+  // compressed axis. This allows us to create indices for a compressed matrix
+  // and de-batch them later (two kernels). Otherwise we would have to create
+  // indices for each batch individually requiring n_batch kernels. For csr/bsr,
+  // we already have the batch dim adjacent to the compressed axis and can
+  // flatten them together. For csc/bsc, we need to transpose first.
+  // For BSR/CSR (b, r, c) -> (b*r, c)
+  // For BSC/CSC (b, c, r) -> (r, b*c)
+  AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+      target_layout,
+      "dense_to_sparse_compressed",
+      [&]() {
+        values = values.flatten(0, 1);
+        mask = mask.flatten(0, 1);
+      },
+      [&]() {
+        values = values.transpose(0, 1).flatten(1, 2);
+        mask = mask.transpose(0, 1).flatten(1, 2);
+      });
+}
+
+// This function unfolds the compressed indices of a compressed sparse matrix
+// into a batched compressed sparse tensor.
+// This is analogous to an unflatten-like operation:
+// unflatten(0, {b, r}) for csr/bsr with input shape (r*b, c)
+//          (output shape (b, r, c))
+// unflatten(1, {b, c}).transpose(0,1) for csc/bsc with input shape (r, c*b)
+//          (output shape (r, b, c) unflatten, (b, r, c) unflatten + transpose)
+// This only operates on the compressed indices as the plain indices and values
+// can be manipulated as described above without special handling.
+// It is a prerequisite for the conversion that the sparsity pattern is sane for
+// the batched shape. That is each batch has the same number of nonzero
+// elements.
+Tensor compressed_to_batched_compressed_indices(
+    const Tensor& compressed_in,
+    const int64_t& n_batch,
+    bool out_int32) {
+  auto n_compressed_per_batch = (compressed_in.size(0) - 1) / n_batch;
+  ScalarType out_type = out_int32 ? ScalarType::Int : ScalarType::Long;
+  auto batched_out = at::zeros(
+      {n_batch, n_compressed_per_batch + 1},
+      compressed_in.options().dtype(out_type));
+
+  // If the compressed dimension has length zero there is 1 element in each
+  // batch and it is zero we already have this result formed
+  if (n_compressed_per_batch > 0) {
+    auto trailing_slice =
+        compressed_in.slice(0, 1, c10::nullopt, 1).reshape({n_batch, -1});
+    auto offsets = compressed_in.slice(0, 0, -1, n_compressed_per_batch)
+                       .reshape({n_batch, -1});
+    batched_out.narrow(-1, 1, n_compressed_per_batch)
+        .copy_(trailing_slice - offsets);
```

Comment:
@nikitaved I sort of minced your suggestions together with what was already here. 

Yes, it could go in a single statement, but it is very difficult to parse and the two intermediate variables `trailing_slice` and `offsets` make it much clearer.

Reply:
And you decided against in-placeness, right? We could probably give a name to the difference?

Reply:
Sorry, yes. The `copy_` is fine however we can't do the subtraction in place as both factors are views on the same storage, and that sub does in fact broadcast `(n_batch x n_compressed_per_batch)` - (n_batch x  1)

Reply:
Yeah, missed that. Good catch!

Reply:
Alternatively, we could copy first, then do in-place diff :)